src/middlewared/middlewared/plugins/auth.py

```diff
@@ -388,124 +383,511 @@ def get_token_for_shell_application(self, token_id, origin):
+                            },
+                            'error': resp['pam_response']['reason']
+                        }, False)
+
+            case AuthMech.API_KEY_PLAIN:
+                # API key that we receive over wire is concatenation of the
+                # datastore `id` of the particular key with the key itself,
+                # delimited by a dash. <id>-<key>.
+                resp = await self.get_login_user(
+                    app,
+                    data['username'],
+                    data['api_key'],
+                    mechanism
+                )
+                if resp['pam_response']['code'] == pam.PAM_AUTHINFO_UNAVAIL:
+                    # This is a special error code that means we need to
+                    # etc.generate because we somehow got garbage in the file.
+                    # It should not happen, but we must try to recover.
+
+                    self.logger.warning('API key backend has errors that require regenerating its file.')
+                    await self.middleware.call('etc.generate', 'pam_middleware')
+
+                    # We've exhausted steps we can take, so we'll take the
+                    # response to second request as authoritative
+                    resp = await self.get_login_user(
+                        app,
+                        data['username'],
+                        data['api_key'],
+                        mechanism
+                    )
+
+                # Retrieve the API key here so that we can upgrade the underlying
+                # hash type and iterations if needed (since we have plain-text).
+                # We also need the key info so that we can generate a useful
+                # audit entry in case of failure.
+                try:
+                    key_id = int(data['api_key'].split('-')[0])
+                    key = await self.middleware.call(
+                        'api_key.query', [['id', '=', key_id]],
+                        {'get': True, 'select': ['id', 'name', 'keyhash', 'expired']}
+                    )
+                    thehash = key.pop('keyhash')
+                except Exception:
+                    key = None
+
+                if resp['pam_response']['code'] == pam.PAM_CRED_EXPIRED:
+                    # Give more precise reason for login failure for audit trails
+                    # because we need to differentiate between key and account
+                    # being expired.
+                    resp['pam_response']['reason'] = 'Api key is expired.'
+
+                if resp['pam_response']['code'] == pam.PAM_SUCCESS:
+                    if thehash.startswith('$pbkdf2-sha256'):
+                        # Legacy API key with insufficient iterations. Since we
+                        # know that the plain-text we have here is correct, we can
+                        # use it to update the hash in backend.
+                        await self.middleware.call('api_key.update_hash', data['api_key'])
+
```

Comment:
IIRC the plan was to auto-revoke an API_KEY that was used on a non-private transport.  I didn't notice this being done anywhere.  Would here be a good location?

Reply:
Yes it would, but in a separate commit.